File: qt/gen_qlistview.cpp
// +build ignore

#include <QAbstractItemView>
#include <QAbstractScrollArea>
#include <QDragLeaveEvent>
#include <QDragMoveEvent>
#include <QDropEvent>
#include <QEvent>
#include <QFrame>
#include <QItemSelection>
#include <QList>
#include <QListView>
#include <QMetaObject>
#include <QModelIndex>
#include <QMouseEvent>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QPoint>
#include <QRect>
#include <QRegion>
#include <QResizeEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionViewItem>
#include <QTimerEvent>
#include <QWheelEvent>
#include <QWidget>
#include <qlistview.h>
#include "gen_qlistview.h"
class MiqtVirtualQListView : public virtual QListView {
public:
MiqtVirtualQListView(QWidget* parent): QListView(parent) {};
MiqtVirtualQListView(): QListView() {};

virtual ~MiqtVirtualQListView() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QListView::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QListView_MetaObject(const_cast<MiqtVirtualQListView*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QListView::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QListView::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QListView_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QListView::qt_metacast(param1);

	}
};
QListView* QListView_new(QWidget* parent) {
return new MiqtVirtualQListView(parent);
}
QListView* QListView_new2() {
return new MiqtVirtualQListView();
}
void QListView_virtbase(QListView* src
, QAbstractItemView** outptr_QAbstractItemView
) {
*outptr_QAbstractItemView = static_cast<QAbstractItemView*>(src);
}
QMetaObject* QListView_MetaObject(const QListView* self) {
	return (QMetaObject*) self->metaObject();
}
void* QListView_Metacast(QListView* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QListView_Tr(const char* s) {
	QString _ret = QListView::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QListView_SetMovement(QListView* self, Movement movement) {
	self->setMovement(movement);
}
Movement QListView_Movement(const QListView* self) {
	return self->movement();
}
void QListView_SetFlow(QListView* self, Flow flow) {
	self->setFlow(flow);
}
Flow QListView_Flow(const QListView* self) {
	return self->flow();
}
void QListView_SetWrapping(QListView* self, bool enable) {
	self->setWrapping(enable);
}
bool QListView_IsWrapping(const QListView* self) {
	return self->isWrapping();
}
void QListView_SetResizeMode(QListView* self, ResizeMode mode) {
	self->setResizeMode(mode);
}
ResizeMode QListView_ResizeMode(const QListView* self) {
	return self->resizeMode();
}
void QListView_SetLayoutMode(QListView* self, LayoutMode mode) {
	self->setLayoutMode(mode);
}
LayoutMode QListView_LayoutMode(const QListView* self) {
	return self->layoutMode();
}
void QListView_SetSpacing(QListView* self, int space) {
	self->setSpacing(static_cast<int>(space));
}
int QListView_Spacing(const QListView* self) {
	return self->spacing();
}
void QListView_SetBatchSize(QListView* self, int batchSize) {
	self->setBatchSize(static_cast<int>(batchSize));
}
int QListView_BatchSize(const QListView* self) {
	return self->batchSize();
}
void QListView_SetGridSize(QListView* self, QSize* size) {
	self->setGridSize(*size);
}
QSize* QListView_GridSize(const QListView* self) {
	return new QSize(self->gridSize());
}
void QListView_SetViewMode(QListView* self, ViewMode mode) {
	self->setViewMode(mode);
}
ViewMode QListView_ViewMode(const QListView* self) {
	return self->viewMode();
}
void QListView_ClearPropertyFlags(QListView* self) {
	self->clearPropertyFlags();
}
bool QListView_IsRowHidden(const QListView* self, int row) {
	return self->isRowHidden(static_cast<int>(row));
}
void QListView_SetRowHidden(QListView* self, int row, bool hide) {
	self->setRowHidden(static_cast<int>(row), hide);
}
void QListView_SetModelColumn(QListView* self, int column) {
	self->setModelColumn(static_cast<int>(column));
}
int QListView_ModelColumn(const QListView* self) {
	return self->modelColumn();
}
void QListView_SetUniformItemSizes(QListView* self, bool enable) {
	self->setUniformItemSizes(enable);
}
bool QListView_UniformItemSizes(const QListView* self) {
	return self->uniformItemSizes();
}
void QListView_SetWordWrap(QListView* self, bool on) {
	self->setWordWrap(on);
}
bool QListView_WordWrap(const QListView* self) {
	return self->wordWrap();
}
void QListView_SetSelectionRectVisible(QListView* self, bool show) {
	self->setSelectionRectVisible(show);
}
bool QListView_IsSelectionRectVisible(const QListView* self) {
	return self->isSelectionRectVisible();
}
void QListView_SetItemAlignment(QListView* self, int alignment) {
	self->setItemAlignment(static_cast<Qt::Alignment>(alignment));
}
int QListView_ItemAlignment(const QListView* self) {
	Qt::Alignment _ret = self->itemAlignment();
	return static_cast<int>(_ret);
}
QRect* QListView_VisualRect(const QListView* self, QModelIndex* index) {
	return new QRect(self->visualRect(*index));
}
void QListView_ScrollTo(QListView* self, QModelIndex* index, ScrollHint hint) {
	self->scrollTo(*index, hint);
}
QModelIndex* QListView_IndexAt(const QListView* self, QPoint* p) {
	return new QModelIndex(self->indexAt(*p));
}
void QListView_DoItemsLayout(QListView* self) {
	self->doItemsLayout();
}
void QListView_Reset(QListView* self) {
	self->reset();
}
void QListView_SetRootIndex(QListView* self, QModelIndex* index) {
	self->setRootIndex(*index);
}
void QListView_IndexesMoved(QListView* self, struct miqt_array /* of QModelIndex* */  indexes) {
	QModelIndexList indexes_QList;
	indexes_QList.reserve(indexes.len);
	QModelIndex** indexes_arr = static_cast<QModelIndex**>(indexes.data);
	for(size_t i = 0; i < indexes.len; ++i) {
		indexes_QList.push_back(*(indexes_arr[i]));
	}
	self->indexesMoved(indexes_QList);
}
void QListView_connect_IndexesMoved(QListView* self, intptr_t slot) {
	MiqtVirtualQListView::connect(self, static_cast<void (QListView::*)(const QModelIndexList&)>(&QListView::indexesMoved), self, [=](const QModelIndexList& indexes) {
		const QModelIndexList& indexes_ret = indexes;
		// Convert QList<> from C++ memory to manually-managed C memory
		QModelIndex** indexes_arr = static_cast<QModelIndex**>(malloc(sizeof(QModelIndex*) * indexes_ret.length()));
		for (size_t i = 0, e = indexes_ret.length(); i < e; ++i) {
			indexes_arr[i] = new QModelIndex(indexes_ret[i]);
		}
		struct miqt_array indexes_out;
		indexes_out.len = indexes_ret.length();
		indexes_out.data = static_cast<void*>(indexes_arr);
		struct miqt_array /* of QModelIndex* */  sigval1 = indexes_out;
		miqt_exec_callback_QListView_IndexesMoved(slot, sigval1);
	});
}
struct miqt_string QListView_Tr2(const char* s, const char* c) {
	QString _ret = QListView::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QListView_Tr3(const char* s, const char* c, int n) {
	QString _ret = QListView::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QListView_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQListView*>( (QListView*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QListView_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQListView*)(self) )->virtualbase_MetaObject();
}
void QListView_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQListView*>( (QListView*)(self) )->handle__Metacast = slot;
}
void* QListView_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQListView*)(self) )->virtualbase_Metacast(param1);
}
void QListView_Delete(QListView* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQListView*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qlistview.go
package qt

import (
	"unsafe"
)

type QListView__Movement int

const (
	QListView__Static QListView__Movement = 0
	QListView__Free   QListView__Movement = 1
	QListView__Snap   QListView__Movement = 2
)

type QListView__Flow int

const (
	QListView__LeftToRight QListView__Flow = 0
	QListView__TopToBottom QListView__Flow = 1
)

type QListView__ResizeMode int

const (
	QListView__Fixed  QListView__ResizeMode = 0
	QListView__Adjust QListView__ResizeMode = 1
)

type QListView__LayoutMode int

const (
	QListView__SinglePass QListView__LayoutMode = 0
	QListView__Batched    QListView__LayoutMode = 1
)

type QListView__ViewMode int

const (
	QListView__ListMode QListView__ViewMode = 0
	QListView__IconMode QListView__ViewMode = 1
)

type QListView struct {
	h          uintptr
	isSubclass bool
}

// NewQListView constructs a new QListView object.
func NewQListView(parent *QWidget) *QListView {
	ret := newQListView(QListView_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQListView2 constructs a new QListView object.
func NewQListView2() *QListView {
	ret := newQListView(QListView_new2())
	ret.isSubclass = true
	return ret
}

func (this *QListView) MetaObject() *QMetaObject {
	return newQMetaObject(QListView_MetaObject(this.h))
}

func (this *QListView) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QListView_Metacast(this.h, param1_Cstring))
}

func QListView_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QListView_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QListView) SetMovement(movement Movement) {
	QListView_SetMovement(this.h, movement)
}

func (this *QListView) Movement() Movement {
	xxxxxxxxx
}

func (this *QListView) SetFlow(flow Flow) {
	QListView_SetFlow(this.h, flow)
}

func (this *QListView) Flow() Flow {
	xxxxxxxxx
}

func (this *QListView) SetWrapping(enable bool) {
	QListView_SetWrapping(this.h, (bool)(enable))
}

func (this *QListView) IsWrapping() bool {
	return (bool)(QListView_IsWrapping(this.h))
}

func (this *QListView) SetResizeMode(mode ResizeMode) {
	QListView_SetResizeMode(this.h, mode)
}

func (this *QListView) ResizeMode() ResizeMode {
	xxxxxxxxx
}

func (this *QListView) SetLayoutMode(mode LayoutMode) {
	QListView_SetLayoutMode(this.h, mode)
}

func (this *QListView) LayoutMode() LayoutMode {
	xxxxxxxxx
}

func (this *QListView) SetSpacing(space int) {
	QListView_SetSpacing(this.h, (int)(space))
}

func (this *QListView) Spacing() int {
	return (int)(QListView_Spacing(this.h))
}

func (this *QListView) SetBatchSize(batchSize int) {
	QListView_SetBatchSize(this.h, (int)(batchSize))
}

func (this *QListView) BatchSize() int {
	return (int)(QListView_BatchSize(this.h))
}

func (this *QListView) SetGridSize(size *QSize) {
	QListView_SetGridSize(this.h, size.cPointer())
}

func (this *QListView) GridSize() *QSize {
	_goptr := newQSize(QListView_GridSize(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QListView) SetViewMode(mode ViewMode) {
	QListView_SetViewMode(this.h, mode)
}

func (this *QListView) ViewMode() ViewMode {
	xxxxxxxxx
}

func (this *QListView) ClearPropertyFlags() {
	QListView_ClearPropertyFlags(this.h)
}

func (this *QListView) IsRowHidden(row int) bool {
	return (bool)(QListView_IsRowHidden(this.h, (int)(row)))
}

func (this *QListView) SetRowHidden(row int, hide bool) {
	QListView_SetRowHidden(this.h, (int)(row), (bool)(hide))
}

func (this *QListView) SetModelColumn(column int) {
	QListView_SetModelColumn(this.h, (int)(column))
}

func (this *QListView) ModelColumn() int {
	return (int)(QListView_ModelColumn(this.h))
}

func (this *QListView) SetUniformItemSizes(enable bool) {
	QListView_SetUniformItemSizes(this.h, (bool)(enable))
}

func (this *QListView) UniformItemSizes() bool {
	return (bool)(QListView_UniformItemSizes(this.h))
}

func (this *QListView) SetWordWrap(on bool) {
	QListView_SetWordWrap(this.h, (bool)(on))
}

func (this *QListView) WordWrap() bool {
	return (bool)(QListView_WordWrap(this.h))
}

func (this *QListView) SetSelectionRectVisible(show bool) {
	QListView_SetSelectionRectVisible(this.h, (bool)(show))
}

func (this *QListView) IsSelectionRectVisible() bool {
	return (bool)(QListView_IsSelectionRectVisible(this.h))
}

func (this *QListView) SetItemAlignment(alignment AlignmentFlag) {
	QListView_SetItemAlignment(this.h, (int)(alignment))
}

func (this *QListView) ItemAlignment() AlignmentFlag {
	return (AlignmentFlag)(QListView_ItemAlignment(this.h))
}

func (this *QListView) VisualRect(index *QModelIndex) *QRect {
	_goptr := newQRect(QListView_VisualRect(this.h, index.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QListView) ScrollTo(index *QModelIndex, hint ScrollHint) {
	QListView_ScrollTo(this.h, index.cPointer(), hint)
}

func (this *QListView) IndexAt(p *QPoint) *QModelIndex {
	_goptr := newQModelIndex(QListView_IndexAt(this.h, p.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QListView) DoItemsLayout() {
	QListView_DoItemsLayout(this.h)
}

func (this *QListView) Reset() {
	QListView_Reset(this.h)
}

func (this *QListView) SetRootIndex(index *QModelIndex) {
	QListView_SetRootIndex(this.h, index.cPointer())
}

func (this *QListView) IndexesMoved(indexes []QModelIndex) {
	indexes_CArray := (*[0xffff]*QModelIndex)(malloc(size_t(8 * len(indexes))))
	defer free(unsafe.Pointer(indexes_CArray))
	for i := range indexes {
		indexes_CArray[i] = indexes[i].cPointer()
	}
	indexes_ma := struct_miqt_array{len: size_t(len(indexes)), data: unsafe.Pointer(indexes_CArray)}
	QListView_IndexesMoved(this.h, indexes_ma)
}

func (this *QListView) OnIndexesMoved(slot func(indexes []QModelIndex)) {
	QListView_connect_IndexesMoved(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QListView_IndexesMoved
func miqt_exec_callback_QListView_IndexesMoved(cb intptr_t, indexes struct_miqt_array) {
	gofunc, ok := cgo.Handle(cb).Value().(func(indexes []QModelIndex))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	var indexes_ma struct_miqt_array = indexes
	indexes_ret := make([]QModelIndex, int(indexes_ma.len))
	indexes_outCast := (*[0xffff]*QModelIndex)(unsafe.Pointer(indexes_ma.data)) // hey ya
	for i := 0; i < int(indexes_ma.len); i++ {
		indexes_lv_goptr := newQModelIndex(indexes_outCast[i])
		indexes_lv_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
		indexes_ret[i] = *indexes_lv_goptr
	}
	slotval1 := indexes_ret

	gofunc(slotval1)
}

func QListView_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QListView_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QListView_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QListView_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QListView) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QListView_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QListView) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QListView_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QListView_MetaObject
func miqt_exec_callback_QListView_MetaObject(self QListView, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QListView{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QListView) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QListView_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QListView) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QListView_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QListView_Metacast
func miqt_exec_callback_QListView_Metacast(self QListView, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QListView{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
